### ISSUE TITLE
Reduce redundancy in JoinSplit gadget implementation

### ISSUE THREAD

**Issue by daira** [zcash/zcash#824]
https://github.com/zcash/zcash/blob/zc.v0.11.2.latest/src/zerocash/zerocash_pour_gadget.tcc

A substantial proportion of the code in this file seems like it should not need to exist. For example: the code in ``zerocash_pour_input_map`` has to match that after ``/* do the multipacking */`` on line 61. Why is this done in two places? If they didn't match then there might be a security bug.

**Comment by ebfull** [comment]
The witness map function is used to pack the verifier inputs (bit strings) into field elements for verification of a proof. If they didn't match for whatever reason, all proofs would be highly unlikely to verify. Basic code review can ensure that they match.

I've also made an effort to refactor and clean up the code so it's much easier to review in this respect. Completely removing the redundancy is not really practical, though, thanks to a lack of expressiveness in C++. I was able to come up with a solution that used twice the amount of code we currently do. It got to the point where the abstraction was more likely to be flawed than the previous code. We *could* do it succinctly with C-style macros (I think?) but I want to avoid those in our circuit code for obvious reasons.

Thus, I'm opting for the simpler and more elegant approach and look forward to seeing what others think. I nominate to have this removed from this milestone. (I think @daira can make this decision unilaterally to avoid bugging the team.)
